effectful-th/src/Effectful/TH.hs

```diff
@@ -0,0 +1,243 @@
+    Left  esName    -> pure (esName, Nothing)
+    Right monadName -> (, Just monadName) <$> newName "es"
+
+  let fnName = mkName . toSmartConName $ nameBase name
+  fnArgs <- traverse (const $ newName "x") actionParams
+
+  let esVar = VarT esName
+
+      substM :: Type -> Type
+      substM = case maybeMonadName of
+        Just m  -> applySubstitution . Map.singleton m $ ConT ''Eff `AppT` esVar
+        Nothing -> id
+
+      (origActionVars, actionCtx) = extractCtx typ
+      actionVars = case maybeMonadName of
+        Just m  -> filter ((m /=) . tvName) origActionVars
+                ++ [kindedTVSpecified esName $ ListT `AppT` ConT ''Effect]
+        Nothing -> origActionVars
+
+#if MIN_VERSION_template_haskell(2,17,0)
+  -- In GHC >= 9.0 it's possible to generate the following body:
+  --
+  -- e x1 .. xN = send (E @ty1 .. @tyN x1 .. xN)
+  --
+  -- because specificities of constructor variables are exposed.
+  --
+  -- This allows to generate functions for such effects:
+  --
+  -- type family F ty :: Type
+  -- data AmbEff :: Effect where
+  --   AmbEff :: Int -> AmbEff m (F ty)
+  --
+  -- Sadly the version for GHC < 9 will not compile due to ambiguity error.
+  let fnBody =
+        let tvFlag = \case
+              PlainTV  _ flag   -> flag
+              KindedTV _ flag _ -> flag
+
+            tyApps = (`mapMaybe` origActionVars) $ \v -> case tvFlag v of
+              InferredSpec  -> Nothing
+              SpecifiedSpec -> Just $ if maybeMonadName == Just (tvName v)
+                                      then ConT ''Eff `AppT` esVar
+                                      else VarT (tvName v)
+
+            effCon = if makeSig
+              then foldl' AppTypeE (ConE name) tyApps
+              else                  ConE name
+        in VarE 'send `AppE` foldl' (\f -> AppE f . VarE) effCon fnArgs
+#else
+  -- In GHC < 9.0, generate the following body:
+  --
+  -- e :: E v1 .. vN :> es => x1 -> .. -> xK -> E v1 .. vN (Eff es) r
+  -- e x1 .. xK = send (E x1 .. xN :: E v1 .. vK (Eff es) r)
+  let fnBody =
+        let effOp  = foldl' (\f -> AppE f . VarE) (ConE name) fnArgs
+            effSig = effTy `AppT` (ConT ''Eff `AppT` esVar) `AppT` substM resTy
+        in if makeSig
+           then VarE 'send `AppE` SigE effOp effSig
+           else VarE 'send `AppE`      effOp
+#endif
+  let fnSig = ForallT actionVars
+        (UInfixT effTy ''(:>) esVar : actionCtx)
+        (makeTyp esVar substM resTy actionParams)
+
+  let rest = FunD fnName [Clause (VarP <$> fnArgs) (NormalB fnBody) []]
+           : maybeToList ((`InfixD` name) <$> fixity)
+  (++ rest) <$> withHaddock name [SigD fnName fnSig | makeSig]
+
+----------------------------------------
+-- Helpers
+
+toSmartConName :: String -> String
+toSmartConName = \case
+  (':' : xs) -> xs
+  (x : xs)   -> toLower x : xs
+  _          -> error "empty constructor name"
+
+extractCtx :: Type -> ([TyVarBndrSpec], Cxt)
+extractCtx = \case
+  ForallT vars ctx _ -> (vars, ctx)
```

Comment:
Is this not recursive because the reified type will definitely be normalized?

Reply:
GADT constructors do not allow nested foralls: https://gitlab.haskell.org/ghc/ghc/-/issues/18191